repose-aggregator/src/docs/asciidoc/recipes/functional-test-framework.adoc

```diff
@@ -0,0 +1,163 @@
+
+This framework is based on the http://spockframework.org/[Spock framework].
+While there are features of this framework that are can be used without Spock, it is expected that tests utilizing this framework will be Spock tests.
+
+== Download
+
+The Repose functional test framework is published to https://maven.research.rackspacecloud.com/[an instance of Nexus hosted by Rackspace].
+
+It is highly recommended that a build system be used to manage this library.
+The following snippets provide the necessary additions to download this framework with some popular build systems.
+
+=== Maven
+
+[source,xml]
+.pom.xml
+----
+<repositories>
+    <repository>
+        <id>public.maven.research.rackspace.com</id>
+        <name>Rackspace Research</name>
+        <url>https://maven.research.rackspacecloud.com/content/groups/aggregate</url>
+    </repository>
+</repositories>
+
+<dependencies>
+    <dependency>
+      <groupId>org.openrepose</groupId>
+      <artifactId>functional-test-framework</artifactId>
+      <version>${version}</version>
+    </dependency>
+</dependencies>
+----
+
+=== Gradle
+
+[source,groovy]
+.build.gradle
+----
+repositories {
+    maven {
+        url "https://maven.research.rackspacecloud.com/content/groups/aggregate/"
+    }
+}
+
+dependencies {
+    compile "org.openrepose:functional-test-framework:${version}"
+}
+----
+
+== Setup
+
+There are two requirements that must be satisfied for this framework.
+
+The first requirement is that a `repose-test.properties` file is on the classpath when tests are run.
+The `repose-test.properties` file tells that framework what version of Repose, directories, host, and ports to use.
+It should be in the https://en.wikipedia.org/wiki/.properties#Format[Java properties file format] like the following example:
+
+----
+test.root.directory=/default/test/root # <1>
+
+repose.version=${version} # <2>
+repose.container=valve # <3>
+repose.home=/default/repose/home # <4>
+repose.config.directory=/configs # <5>
+repose.config.templates=/default/config/templates # <6>
+repose.log.name=/logs/repose.log # <7>
+repose.lint.log.name=/logs/repose-lint.log # <8>
+repose.log.pattern=/logs/repose_%d{yyyy-MM-dd_HHmmss}.log # <9>
+
+repose.jar=/default/valve/location # <10>
+repose.lint.jar=/default/lint/location # <11>
+
+repose.root.war=/default/root/war/location # <12>
+glassfish.jar=/default/glassfish/location # <13>
+tomcat.jar=/default/tomcat/location # <14>
+mocks.war=/default/mocks/war/location # <15>
+
+target.hostname=localhost # <16>
+
+port.finder.port.start=10000 # <17>
+port.finder.port.max=65535 # <18>
+port.finder.port.range=100 # <19>
+port.finder.property.name.worker.id=org.gradle.test.worker # <20>
+----
+<1> The absolute path of the root directory to run tests under.
+The user running the tests must have permission to read and write in this directory.
+<2> The version of Repose.
+<3> The servlet container to run Repose in.
```

Comment:
Weirdly, I put a path in the default properties file. So that's weird. I'll stick valve in there.

As for what values work, I'm not sure what you mean there. The `TestProperties` class doesn't care what the value is (no validation). `ReposeValveTest` does, but it seems to also take the value `multinode`. Of course, there is no requirement that `ReposeValveTest` be used anyway.